onnxscript/function_libs/torch_aten/ops/nn.py

```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
     raise NotImplementedError()
 
 
+@onnxscript.script()
+def _causal_attention_mask(query: TFloat, key: TFloat) -> TFloat:
+    # attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
+    # attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool
+    target_length = op.Shape(query)[-2:-1]
+    source_length = op.Shape(key)[-2:-1]
+    # attn_mask = torch.ones(L, S) := {
+    size = op.Concat(target_length, source_length, axis=0)
+    attn_mask = op.Expand(1.0, size)
+    # }
+    attn_mask = op.Trilu(attn_mask, upper=0)
+    # The causal mask has 0s in the lower triangle and -inf in the upper triangle.
+    attn_mask = op.Where(op.Equal(attn_mask, 0.0), op.Constant(value_float=-float("inf")), 0.0)
+    return attn_mask
+
+
+@onnxscript.script()
+def _attention_scale(query: TFloat) -> TFloat:
+    # scale_factor = 1 / math.sqrt(Q.size(-1))
```

Comment:
nit: Q, K, V are pretty obvious, but wonder if it is still better to just write as `query` in comment

Reply:
Done